application/apps/indexer/sources/src/serial/serialport.rs

```diff
@@ -1,11 +1,46 @@
 use crate::factory::SerialTransportConfig;
-// use crate::serial::sde;
+use crate::serial::sde;
 use crate::ByteSource;
 use crate::{Error as SourceError, ReloadInfo, SourceFilter};
 use async_trait::async_trait;
 use buf_redux::Buffer;
-use tokio::io::{AsyncBufReadExt, BufReader};
+use bytes::{BufMut, BytesMut};
+use futures::stream::{SplitSink, SplitStream, StreamExt};
+use futures::SinkExt;
+use std::{io, str};
 use tokio_serial::{DataBits, FlowControl, Parity, SerialPortBuilderExt, SerialStream, StopBits};
+use tokio_util::codec::{Decoder, Encoder, Framed};
+
+struct LineCodec;
+
+impl Decoder for LineCodec {
+    type Item = String;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        match &src.iter().position(|b| *b == b'\n') {
+            Some(n) => match str::from_utf8(&src.split_to(n + 1)) {
+                Ok(s) => Ok(Some(s.to_string())),
+                Err(err) => Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("Failed to format string: {}", err),
+                )),
+            },
+            None => Ok(None),
+        }
+    }
+}
+
+impl Encoder<String> for LineCodec {
+    type Error = io::Error;
+
+    fn encode(&mut self, item: String, dst: &mut BytesMut) -> Result<(), Self::Error> {
+        dst.reserve(item.len() + 1);
+        dst.put(item.as_bytes());
+        dst.put_u8(b'\n');
+        Ok(())
+    }
```

Comment:
I am not sure, but here you get as input `String` and after you "put" this string as bytes into buffer.
But... bellow in `income` income you gets bytes and convert it into string... Looks a little bit wierd. In general it looks like it:
- FE -> Rust:  sends bytes
- Rust (income method): convert bytes to string
- Rust (income method): send string into `write_stream`
- Rust (LineCodec / encode): convert string to bytes and write into buffer

If we are writing bytes into buffer, does it make sense to convert it into string?